api/logs/v1/labels_enforcer.go

```diff
@@ -67,7 +77,24 @@ func enforceValues(lm []*labels.Matcher, v url.Values) (values string, err error
 	expr.Walk(func(expr interface{}) {
 		switch le := expr.(type) {
 		case *logqlv2.StreamMatcherExpr:
-			le.AppendMatchers(lm)
+			matchers := make([]*labels.Matcher, 0)
+			matchersMap := make(map[string]*labels.Matcher)
+			for _, em := range le.Matchers() {
+				matchersMap[em.Name] = em
+			}
+
+			for _, m := range lm {
+				matcher := matchersMap[m.Name]
+				if matcher == nil || !m.Matches(matcher.Value) {
+					matchersMap[m.Name] = m
+				}
+			}
+
+			for _, m := range matchersMap {
+				matchers = append(matchers, m)
+			}
```

Comment:
This is very cool but hard to grasp without consulting the tests. I suggest to offload this to a custom functions and provide some inline docs per loop. WDYT?

Reply:
sure let me see what I can do